### ISSUE TITLE
Inconsistency in mixed date/int arithmetics

### ISSUE THREAD

**Issue by abalkin** [pola-rs/polars#4574]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of polars.

### Issue Description

Adding an integer to a date series behaves differently in select context and outside.

### Reproducible Example

```python
>>> from datetime import date
>>> import polars as pl
>>> df = pl.DataFrame({'date': pl.date_range(date(2001, 1, 1), date(2001, 1, 5), '1d')})
>>> df.select( pl.col('date') + 1 )
shape: (5, 1)
┌───────┐
│ date  │
│ ---   │
│ i32   │
╞═══════╡
│ 11324 │
├╌╌╌╌╌╌╌┤
│ 11325 │
├╌╌╌╌╌╌╌┤
│ 11326 │
├╌╌╌╌╌╌╌┤
│ 11327 │
├╌╌╌╌╌╌╌┤
│ 11328 │
└───────┘

>>> df['date'] + 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/a13116q/asu/.build/dev/lib/python3.9/site-packages/polars/internals/series/series.py", line 408, in __add__
    return self._arithmetic(other, "add", "add_<>")
  File "/Users/a13116q/asu/.build/dev/lib/python3.9/site-packages/polars/internals/series/series.py", line 396, in _arithmetic
    other = maybe_cast(other, self.dtype, self.time_unit)
  File "/Users/a13116q/asu/.build/dev/lib/python3.9/site-packages/polars/datatypes.py", line 540, in maybe_cast
    el = py_type(el)
TypeError: function missing required argument 'month' (pos 2)
```

### Expected Behavior

I expect adding an integer to a date series to add that many days to each element.  The `.select()` behavior reported above is correct, but the return type should be `pl.Date`.

```python
>>> df.select( pl.col('date') + 1 )
shape: (5, 1)
┌────────────┐
│ date       │
│ ---        │
│ date       │
╞════════════╡
│ 2001-01-02 │
├╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2001-01-03 │
├╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2001-01-04 │
├╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2001-01-05 │
├╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2001-01-06 │
└────────────┘
``` 

### Installed Versions

<details>

>>> pl.show_versions()
---Version info---
Polars: 0.14.5
Index type: UInt32
Platform: macOS-12.5-arm64-arm-64bit
Python: 3.9.7 (default, Aug 12 2022, 09:02:06)
[Clang 13.1.6 (clang-1316.0.21.2.5)]
---Optional dependencies---
pyarrow: 9.0.0
pandas: 1.4.3
numpy: 1.22.4
fsspec: 2022.7.1
connectorx: 0.3.0
xlsx2csv: <not installed>

</details>

**Comment by alexander-beedie** [comment]
Small follow-up; adding a `timedelta` instead of an integer _almost_ works as desired but (incorrectly?) upcasts the result to `Datetime('ns')` instead of maintaining the original `Date` dtype:

```python
df.with_columns( (pl.col('date') + timedelta(days=1)).alias('date_plus_one') )
# Out[5]: 
# shape: (5, 2)
# ┌────────────┬─────────────────────┐
# │ date       ┆ date_plus_one       │
# │ ---        ┆ ---                 │
# │ date       ┆ datetime[ns]        │
# ╞════════════╪═════════════════════╡
# │ 2001-01-01 ┆ 2001-01-02 00:00:00 │
# ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
# │ 2001-01-02 ┆ 2001-01-03 00:00:00 │
# ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
# │ 2001-01-03 ┆ 2001-01-04 00:00:00 │
# ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
# │ 2001-01-04 ┆ 2001-01-05 00:00:00 │
# ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
# │ 2001-01-05 ┆ 2001-01-06 00:00:00 │
# └────────────┴─────────────────────┘
```

**Comment by ritchie46** [comment]
We cannot make that work on integers. The supertype of `date + int64` needs to remain `i64` IMO. 

I have fixed @alexander-beedie's example in #4590
